docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+automerge Javscript library to mutate the document (the `Automerge.change`
+call) and then finally get the change between the original version of the 
+document and the new one. 
+
+Now that we have the change we can make a `PATCH` request to 
+`<radicle-implementation>/projects/<project URN>/metadata/<metadata ID>` with
+the following contents:
+
+```json
+{
+    "changes": {
+        "type": "automerge",
+        "change": "<some base64>
+    }
+}
+```
+
+This endpoint will return an error if the change does not match the schema of
+the object. Otherwise the change will be merged in to the object and announced
+to the network.
+
+## Implementation
+
+### Automerge
+
+It is useful to briefly outline how automerge functions in order for the
+following to be sensible. Everything automerge does is based on merging a log
+of operations. An operation might be something like "create a list under the
+'comments' key of the root object", or "insert the character 'a' after the 
+character inserted by the 2nd change actor 1 made". Every operation has an 
+identifier - which allows statements like "the character inserted by the 2nd
+change actor 1 made" to be precise. This operation ID is the combination of a
+unique identifier for each actor, and an always incrementing sequence number.
+This construction, along with sorting by actor IDs in the case of a tie, allows
+us to place operations in a total order which respects causality, i.e. if I add
+an operation then no operation that I could have observed at the time I made
+the operation will come after it in the log.
+
+Automerge defines a number of operations along with merge semantics for those
+operations. More detail on that can be found in [the implementation](https://github.com/automerge/automerge)
+and in [the paper](https://arxiv.org/abs/1608.03960).
+
+Operations in automerge are transported in batches called "changes". Each 
+change references zero or more changes it depends on via their hash. In this 
+manner automerge is similar to git in that it's a hash linked graph of changes.
+
+Despite all the complexity under the hood, the API of automerge is relatively
+simple. Automerge works in terms of "documents", a document is a single log of
+changes. Every time you modify an automerge document you generate a new entry 
+for the change log. Each change is just some bytes. When you receive changes 
+from other actors you just pass these changes (which, again, are just bytes) to
+automerge to add to the change log. The end result is that you load a bunch of
+binary changes and get back a JSON object.
+
+There are some subtleties around preserving user intent when modifying
+documents, but these are not too onerous.
+
+### Object Trees
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+```
+.
+|--manifest
+|--schema
+|  |--schema.json
+|  |--migrations
```

Comment:
Interesting point. I'll have a think about this. It may be worth moving schemas into their own hash graph and then referencing them in a commit trailer much like the author.